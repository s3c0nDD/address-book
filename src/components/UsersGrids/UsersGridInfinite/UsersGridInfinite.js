import React, { useMemo, memo } from 'react';
import PropTypes from 'prop-types';
import InfiniteScroll from 'react-infinite-scroller';
import { makeStyles } from '@material-ui/core/styles';

import UsersGridLayout, { UsersGridLayoutMessage } from '../UsersGridLayout';

const useStyles = makeStyles(theme => ({
  root: {
    margin: theme.spacing(4)
  },
  catalogEnd: {
    marginTop: theme.spacing(4)
  }
}));

/**
 * Checks props and decide if should not re-render
 * @param {object} prevProps previous props
 * @param {object} nextProps next props
 * @returns {boolean}
 */
const shouldNotRerender = (prevProps, nextProps) => {
  return prevProps.users === nextProps.users
    && prevProps.hideView === nextProps.hideView;
};

/**
 * User grid infinite component
 * @param {object} props component's props
 * @param {boolean} props.hasMore has more users to fetch flag
 * @param {array} props.users users data
 * @param {boolean} props.hideView hide grid view flag
 * @param {function} props.onOpenModal open user modal handler
 * @param {function} props.onLoadMore load more users handler
 */
const UsersGridInfinite = memo(({
  hasMore,
  users,
  hideView,
  onOpenModal,
  onLoadMore
}) => {
  const classes = useStyles();

  const scrollDisplayStyle = useMemo(
    () => (hideView ? { display: 'none' } : {}),
    [hideView]
  );

  return (
    <InfiniteScroll
      className={classes.root}
      style={scrollDisplayStyle}
      loadMore={onLoadMore}
      hasMore={hasMore}
      threshold={500}
    >
      <UsersGridLayout
        users={users}
        onOpenModal={onOpenModal}
      />
      {!hasMore && <UsersGridLayoutMessage
        className={classes.catalogEnd}
        message={'End of users catalog'}
      />}
    </InfiniteScroll>
  );
}, shouldNotRerender);

export default UsersGridInfinite;

UsersGridInfinite.propTypes = {
  hasMore: PropTypes.bool,
  users: PropTypes.array.isRequired,
  hideView: PropTypes.bool,
  onOpenModal: PropTypes.func.isRequired,
  onLoadMore: PropTypes.func.isRequired
};