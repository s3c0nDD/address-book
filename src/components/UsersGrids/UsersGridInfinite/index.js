import UsersGridInfinite from './UsersGridInfinite';

export default UsersGridInfinite;