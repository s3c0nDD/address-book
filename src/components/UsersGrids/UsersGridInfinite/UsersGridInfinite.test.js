import React from 'react';
import { createShallow } from '@material-ui/core/test-utils';

import userMockData from '../../mocks/userMock';
import UsersGridInfinite from './UsersGridInfinite'

describe('UsersGridInfinite', () => {
  let shallow;
  let wrapper;
  let component;
  let openModalMock;
  let loadMoreMock;
  let usersMock = [{ ...userMockData }, { ...userMockData, id: { value: 'foobaz' }}];

  beforeAll(() => {
    shallow = createShallow();
  });

  beforeEach(() => {
    openModalMock = jest.fn();
    loadMoreMock = jest.fn();
    component = (
      <UsersGridInfinite
        hasMore={true}
        users={usersMock}
        hideView={false}
        onOpenModal={openModalMock}
        onLoadMore={loadMoreMock}
      />
    );
    wrapper = shallow(component);
  });

  it('shallowly renders and matches snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});