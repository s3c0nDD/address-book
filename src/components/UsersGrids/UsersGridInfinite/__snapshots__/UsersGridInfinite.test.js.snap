// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersGridInfinite shallowly renders and matches snapshot 1`] = `
<InfiniteScroll
  className="makeStyles-root-1"
  element="div"
  getScrollParent={null}
  hasMore={true}
  initialLoad={true}
  isReverse={false}
  loadMore={[MockFunction]}
  loader={null}
  pageStart={0}
  ref={null}
  style={Object {}}
  threshold={500}
  useCapture={false}
  useWindow={true}
>
  <Memo()
    onOpenModal={[MockFunction]}
    users={
      Array [
        Object {
          "cell": "111-222-333",
          "email": "johndoe123@google.com",
          "id": Object {
            "value": "foobar",
          },
          "location": Object {
            "city": "Warsaw",
            "country": "Poland",
            "postcode": 56789,
            "state": "Masovian",
            "street": Object {
              "name": "Big Street",
              "number": 98,
            },
          },
          "login": Object {
            "username": "johndoe123",
          },
          "name": Object {
            "first": "John",
            "last": "Doe",
            "title": "Mr",
          },
          "phone": "000-111-222",
          "picture": Object {
            "large": "https://wwww.image.com/data/1",
          },
        },
        Object {
          "cell": "111-222-333",
          "email": "johndoe123@google.com",
          "id": Object {
            "value": "foobaz",
          },
          "location": Object {
            "city": "Warsaw",
            "country": "Poland",
            "postcode": 56789,
            "state": "Masovian",
            "street": Object {
              "name": "Big Street",
              "number": 98,
            },
          },
          "login": Object {
            "username": "johndoe123",
          },
          "name": Object {
            "first": "John",
            "last": "Doe",
            "title": "Mr",
          },
          "phone": "000-111-222",
          "picture": Object {
            "large": "https://wwww.image.com/data/1",
          },
        },
      ]
    }
  />
</InfiniteScroll>
`;
